cpp/src/arrow/util/async_generator.h

```diff
@@ -1058,18 +1109,73 @@ class MergedGenerator {
       return source();
     }
 
+    void SignalErrorUnlocked() {
```

Comment:
nit, but it seems inconsistent which ones are explicitly tagged `Unlocked`; it seems all except `PullSource` and `MarkFinalError` assume the mutex is currently held (maybe taking a `const GuardType&` would be more explicit about this precondition)

Reply:
Good idea.  I added the const& and made sure the names were consistent.